### PR TITLE
Add BioGPT

### PR DESCRIPTION
**Pull request by kamalkraj**
# What does this PR do?

Adding BioGPT

Original Implementation and weights - https://github.com/microsoft/BioGPT

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@sgugger @patrickvonplaten

### CODE REVIEW

**Review by younesbelkada — COMMENTED** [review]
Thank you very much for this great contribution and introducing one of the first bio-medical text model to `transformers` ! 
I left a couple of minor comments, overall the PR looks really good to me! The main point being changing the naming of `xxxLMHeadModel` that needs to be changed to `xxxForCausalLM`, I also left a comment about layerdrop and `output_hidden_states`, let me know if this refactoring is possible! 
Another point we should discuss is whether we can use smaller model for integration tests, let me know if you need help on that (no need to train the model)
Great effort on the integration side 💪 We should be very close merging this!

File: docs/source/en/model_doc/biogpt.mdx [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+Tips:
+
+- BioGPT is a model with absolute position embeddings so it’s usually advised to pad the inputs on the right rather than the left.
+- BioGPT was trained with a causal language modeling (CLM) objective and is therefore powerful at predicting the next token in a sequence. Leveraging this feature allows BioGPT to generate syntactically coherent text as it can be observed in the run_generation.py example script.
+- The model can take the past_key_values (for PyTorch) as input, which is the previously computed key/value attention pairs. Using this (past_key_values or past) value prevents the model from re-computing pre-computed values in the context of text generation. For PyTorch, see past_key_values argument of the BioGptLMHeadModel.forward() method for more information on its usage.
+
+This model was contributed by [kamalkraj](<https://huggingface.co/kamalkraj). The original code can be found [here](https://github.com/microsoft/BioGPT).
```
[the remark this answers is not in the document]
Comment (kamalkraj) [reply docs/source/en/model_doc/biogpt.mdx]: 6703338

File: src/transformers/__init__.py [hunk 2]
```diff
@@ -858,6 +859,15 @@
 
     # PyTorch models structure
 
+    _import_structure["models.biogpt"].extend(
+        [
+            "BIOGPT_PRETRAINED_MODEL_ARCHIVE_LIST",
+            "BioGptLMHeadModel",
```
[the remark this answers is not in the document]
Comment (kamalkraj) [reply src/transformers/__init__.py]: 3d07e81

File: src/transformers/models/biogpt/modeling_biogpt.py [hunk 3]
```diff
@@ -0,0 +1,721 @@
...
+        next_decoder_cache = () if use_cache else None
+
+        for idx, decoder_layer in enumerate(self.layers):
+            # add LayerDrop (see https://arxiv.org/abs/1909.11556 for description)
+            if output_hidden_states:
+                all_hidden_states += (hidden_states,)
+            dropout_probability = random.uniform(0, 1)
+            if self.training and (dropout_probability < self.layerdrop):
+                continue
+
```
[the remark this answers is not in the document]
Comment (kamalkraj) [reply src/transformers/models/biogpt/modeling_biogpt.py:552]: Hi @younesbelkada ,

As per my understanding, when `output_hidden_states` is true, we get all hidden states like this
`[embedding_layer_out]` + `[decoder_layer_0_hidden_state, ..., decoder_layer_n_hidden_state]`

`len(all_hidden_states) == 1 + num_hidden_layers`

If we change as you suggested, we won't get the `[embedding_layer_out]`.

Please let me know if I'm missing anything

File: tests/models/biogpt/test_modeling_biogpt.py [hunk 4]
```diff
@@ -0,0 +1,398 @@
...
+
+    def test_biogpt_weight_initialization(self):
+        config_and_inputs = self.model_tester.prepare_config_and_inputs()
+        self.model_tester.create_and_check_biogpt_weight_initialization(*config_and_inputs)
+
+    @slow
+    def test_batch_generation(self):
+        model = BioGptLMHeadModel.from_pretrained("kamalkraj/biogpt")
+        model.to(torch_device)
+        tokenizer = BioGptTokenizer.from_pretrained("kamalkraj/biogpt")
```
[the remark this answers is not in the document]
Comment (kamalkraj) [reply tests/models/biogpt/test_modeling_biogpt.py:308]: I can change this to a randomly initialised small checkpoint. 
But the output from the model will be gibberish, For example
Screenshot 2022-11-26 at 1 30 59 PM

In this way, I don't think we will be able to understand if anything goes wrong.

**Comment by kamalkraj** [comment]
@younesbelkada 
Done changes according to your suggestions.
Thanks for the review

**Comment by younesbelkada** [comment]
Thanks @kamalkraj 
Let me give it another round of review and I'll get back to you

File: docs/source/en/model_doc/biogpt.mdx [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+specific language governing permissions and limitations under the License.
+-->
+
+# BioGPT
+
+## Overview
+
+The BioGPT model was proposed in [BioGPT: generative pre-trained transformer for biomedical text generation and mining
+](https://academic.oup.com/bib/advance-article/doi/10.1093/bib/bbac409/6713511?guestAccessKey=a66d9b5d-4f83-4017-bb52-405815c907b9)  by Renqian Luo, Liai Sun, Yingce Xia, Tao Qin, Sheng Zhang, Hoifung Poon and Tie-Yan Liu. BioGPT is a domain-specific generative pre-trained Transformer language model for biomedical text generation and mining. BioGPT follows the Transformer language model backbone, and is pre-trained on 15M PubMed abstracts from scratch.
```
Comment (younesbelkada) [inline docs/source/en/model_doc/biogpt.mdx:18]:
```suggestion
](https://academic.oup.com/bib/advance-article/doi/10.1093/bib/bbac409/6713511?guestAccessKey=a66d9b5d-4f83-4017-bb52-405815c907b9) by Renqian Luo, Liai Sun, Yingce Xia, Tao Qin, Sheng Zhang, Hoifung Poon and Tie-Yan Liu. BioGPT is a domain-specific generative pre-trained Transformer language model for biomedical text generation and mining. BioGPT follows the Transformer language model backbone, and is pre-trained on 15M PubMed abstracts from scratch.
```

File: src/transformers/models/biogpt/modeling_biogpt.py [hunk 6]
```diff
@@ -0,0 +1,722 @@
...
+class BioGptAttention(nn.Module):
+    """Multi-headed attention from 'Attention Is All You Need' paper"""
+
+    def __init__(
+        self,
+        embed_dim: int,
+        num_heads: int,
+        dropout: float = 0.0,
+        is_decoder: bool = False,
+        bias: bool = True,
```
Comment (younesbelkada) [inline src/transformers/models/biogpt/modeling_biogpt.py:112]: What is the purpose of `bias` here if it is always set to True? I don't think it's an issue here since we have also flagged this on `OPT`, let's just keep this in mind in any case cc @sgugger

File: src/transformers/models/biogpt/modeling_biogpt.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (younesbelkada) [reply src/transformers/models/biogpt/modeling_biogpt.py:552]: thanks for the clarification, I think that you are right here, I am fine with this change. Let's keep it like this

File: tests/models/biogpt/test_modeling_biogpt.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (younesbelkada) [reply tests/models/biogpt/test_modeling_biogpt.py:308]: But even if the model outputs gibberish I'd expect the logits to be different if someone changes the code in the future, so I guess we'll be able to flag it. 
In my opinion I think that it's fine as it is but let's see what others think for final approval - so let's keep this as unresolved for now

File: tests/models/biogpt/test_modeling_biogpt.py [hunk 7]
```diff
@@ -0,0 +1,398 @@
...
+        output_str = tokenizer.decode(output_ids[0], skip_special_tokens=True)
+
+        EXPECTED_OUTPUT_STR = (
+            "COVID-19 is a global pandemic caused by severe acute respiratory syndrome coronavirus 2 (SARS-CoV-2), the"
+            " causative agent of coronavirus disease 2019 (COVID-19), which has spread to more than 200 countries and"
+            " territories, including the United States (US), Canada, Australia, New Zealand, the United Kingdom (UK),"
+            " and the United States of America (USA), as of March 11, 2020, with more than 800,000 confirmed cases and"
+            " more than 800,000 deaths."
+        )
+        self.assertEqual(output_str, EXPECTED_OUTPUT_STR)
```
Comment (younesbelkada) [inline tests/models/biogpt/test_modeling_biogpt.py:398]: Could you confirm you obtained these results under pytorch 1.12 ? Otherwise these tests might fail on our CI runners if the results are different across PyTorch 1.12 and 1.13 for example

**Review by younesbelkada — COMMENTED** [review]
Thanks a lot for the great efforts! I left few minor comments that needs to be double checked, otherwise everything looks great to me!

File: tests/models/biogpt/test_modeling_biogpt.py (hunk 7 above)
Comment (kamalkraj) [reply tests/models/biogpt/test_modeling_biogpt.py:398]: Initial development was done in torch 1.13. 
1.12 also works 

Screenshot 2022-11-28 at 8 28 19 PM

File: docs/source/en/model_doc/biogpt.mdx (hunk 5 above)
Comment (kamalkraj) [reply docs/source/en/model_doc/biogpt.mdx]: done

**Review by younesbelkada — APPROVED** [review]
Thanks so much for your very clean contribution of the first (if I am not mistaken) bio-medical model! 
Clean code & documentation and nice tests 💪🏻 
Leaving now the PR to @sgugger for a final review
Thanks!

File: docs/source/en/model_doc/biogpt.mdx [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+
+Tips:
+
+- BioGPT is a model with absolute position embeddings so it’s usually advised to pad the inputs on the right rather than the left.
+- BioGPT was trained with a causal language modeling (CLM) objective and is therefore powerful at predicting the next token in a sequence. Leveraging this feature allows BioGPT to generate syntactically coherent text as it can be observed in the run_generation.py example script.
+- The model can take the past_key_values (for PyTorch) as input, which is the previously computed key/value attention pairs. Using this (past_key_values or past) value prevents the model from re-computing pre-computed values in the context of text generation. For PyTorch, see past_key_values argument of the BioGptForCausalLM.forward() method for more information on its usage.
```
Comment (sgugger) [inline docs/source/en/model_doc/biogpt.mdx:28]: Use code syntax for past_key_values here (-> \`past_key_values\`)

File: src/transformers/__init__.py [hunk 9]
```diff
@@ -3958,8 +3969,6 @@
             top_k_top_p_filtering,
         )
         from .modeling_utils import PreTrainedModel
-
-        # PyTorch model imports
```
Comment (sgugger) [inline src/transformers/__init__.py:3962]: Should be reverted.

File: src/transformers/__init__.py [hunk 10]
```diff
@@ -4101,6 +4110,14 @@
             BigBirdPegasusModel,
             BigBirdPegasusPreTrainedModel,
         )
+
+        # PyTorch model imports
```
Comment (sgugger) [inline src/transformers/__init__.py:4114]:
```suggestion
```

File: src/transformers/models/auto/modeling_auto.py [hunk 11]
```diff
@@ -305,6 +306,7 @@
         ("bert-generation", "BertGenerationDecoder"),
         ("big_bird", "BigBirdForCausalLM"),
         ("bigbird_pegasus", "BigBirdPegasusForCausalLM"),
+        ("biogpt", "BioGptForCausalLM"),
```
Comment (sgugger) [inline src/transformers/models/auto/modeling_auto.py:309]: You should also add the tokenizer in the auto mappings.

File: src/transformers/models/biogpt/configuration_biogpt.py [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+
+BIOGPT_PRETRAINED_CONFIG_ARCHIVE_MAP = {
+    "kamalkraj/biogpt": "https://huggingface.co/kamalkraj/biogpt/resolve/main/config.json",
+    # See all BioGPT models at https://huggingface.co/models?filter=biogpt
+}
+
+
+class BioGptConfig(PretrainedConfig):
+    r"""
+    This is the configuration class to store the configuration of a [`~BioGptModel`]. It is used to instantiate an
```
Comment (sgugger) [inline src/transformers/models/biogpt/configuration_biogpt.py:31]:
```suggestion
    This is the configuration class to store the configuration of a [`BioGptModel`]. It is used to instantiate an
```

File: src/transformers/models/biogpt/configuration_biogpt.py [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+    [kamalkraj/biogpt](https://huggingface.co/kamalkraj/biogpt) architecture.
+
+    Configuration objects inherit from [`PretrainedConfig`] and can be used to control the model outputs. Read the
+    documentation from [`PretrainedConfig`] for more information.
+
+
+    Args:
+        vocab_size (`int`, *optional*, defaults to 42384):
+            Vocabulary size of the BioGPT model. Defines the number of different tokens that can be represented by the
+            `inputs_ids` passed when calling [`~BioGptModel`].
```
Comment (sgugger) [inline src/transformers/models/biogpt/configuration_biogpt.py:43]:
```suggestion
            `inputs_ids` passed when calling [`BioGptModel`].
```

File: src/transformers/models/biogpt/configuration_biogpt.py [hunk 14]
```diff
@@ -0,0 +1,138 @@
...
+""" BioGPT model configuration"""
+
+from ...configuration_utils import PretrainedConfig
+from ...utils import logging
+
+
+logger = logging.get_logger(__name__)
+
+BIOGPT_PRETRAINED_CONFIG_ARCHIVE_MAP = {
+    "kamalkraj/biogpt": "https://huggingface.co/kamalkraj/biogpt/resolve/main/config.json",
```
Comment (sgugger) [inline src/transformers/models/biogpt/configuration_biogpt.py:24]: Checkpoints will need to be transferred to the microsoft org before we merge.

File: src/transformers/models/biogpt/tokenization_biogpt.py [hunk 15]
```diff
@@ -0,0 +1,373 @@
...
+            token_ids_0 (`List[int]`):
+                List of IDs.
+            token_ids_1 (`List[int]`, *optional*):
+                Optional second list of IDs for sequence pairs.
+
+        Returns:
+            `List[int]`: List of [token type IDs](../glossary#token-type-ids) according to the given sequence(s).
+
+        Creates a mask from the two sequences passed to be used in a sequence-pair classification task. An
+        FAIRSEQ_TRANSFORMER sequence pair mask has the following format:
```
Comment (sgugger) [inline src/transformers/models/biogpt/tokenization_biogpt.py:320]: To remove?

File: src/transformers/models/biogpt/modeling_biogpt.py (hunk 6 above)
Comment (sgugger) [reply src/transformers/models/biogpt/modeling_biogpt.py:112]: Let's leave as is to keep the Copied from.

**Review by sgugger — COMMENTED** [review]
Thanks for adding this model! Once all comments have been resolved, we'll need to move the checkpoints to the microsoft org and adapt all references in this PR :-)

**Comment by kamalkraj** [comment]
@sgugger 
Done changes according to your suggestions.
Thanks for the review

File: src/transformers/models/biogpt/configuration_biogpt.py (hunk 14 above)
Comment (kamalkraj) [reply src/transformers/models/biogpt/configuration_biogpt.py:24]: Let me know once the checkpoints are transferred, I can make the changes.
Thanks

**Comment by younesbelkada** [comment]
Hi @kamalkraj 
The repo has been moved to microsoft: https://huggingface.co/microsoft/biogpt 
Could you please update the PR accordingly? Also it seems that you need to rebase to main
Thanks!

**Comment by kamalkraj** [comment]
@younesbelkada Done the changes 
Thanks

**Comment by younesbelkada** [comment]
thanks @kamalkraj !
It seems that styling tests are failing, could you please run `make fixup`?

**Comment by kamalkraj** [comment]
Hi @younesbelkada ,

The `Add model like runner / Add new model like template tests` install `flake8 6.0` and `ci/circleci: check_code_quality` installs `flask8 5.0.4`. 
Due to this, the above tests are failing. Could you please help me? 

cc @sgugger

**Comment by younesbelkada** [comment]
Thanks so much @kamalkraj !
Let's leave it now to @sgugger to give his review ;) 
Thanks!

**Review by sgugger — APPROVED** [review]
Thanks again for all your work on this!
